### DISCUSSION TITLE
Add mirror command

### DISCUSSION THREAD

**Discussion by JulianOrteil** [ponoor/step-series-support#27] [Ideas]
Hey @kanta 

My application requires two sets of motors with each set driving an axis of an X, Y rail system. Because of this, there is no need for either motor in a set to be configured or run differently than the other. Would you be willing to look into adding a `/mirror` (or other aptly named) command? This command should replicate movement commands and, optionally, configurations and sensor events from one motor to another.

Perhaps the following syntax would be sufficient: `/mirror slaveID masterID shareConfig shareSensors`
- `slaveID` (int): The ID of the motor mirroring the other.
- `masterID` (int): The ID of the motor being mirrored.
- `shareConfig` (bool): Mirror the config (only master->slave). If false, load config for the slave as normal.
- `shareSensors` (bool): Mirror sensor events (only master->slave). The slave should still behave normally if its own sensors fire regardless if this is true.

If any command is sent directly to the slave, the controller should return an error (unless if you think there are valid exceptions, of course).
Let me know what you think!

My Regards,
Jules

**Comment by kanta** [comment]
Hi @JulianOrteil 

Thanks for your idea, it could be a useful command!
At the same time, it would be affect several parts of code and I don't have enough spare time to do that right now. So I would like to　implement this feature as bit longer term Todo, probably late this month or later.

Your definitions of the behavior under the certain conditions are very helpful for implementation, thanks!

**Reply by JulianOrteil** [reply]
@kanta My pleasure. And no worries. This is not something I'll "need" for another month or two, so whenever you have the time, that works for me.

Thanks for your work,
Jules

**Reply by JulianOrteil** [reply]
@kanta Just checking in to see if you've been able to start on this. We're actually getting a noticeable effect in our motion control due to the delay caused by having to send two commands.

Thanks!
Jules

**Reply by kanta** [reply]
@JulianOrteil Sorry, I haven't yet begun the implementation of this.
This change will be affected many parts of the code and need to think about some exceptions.
My idea is

- Behaviors of when a `/mirror` command sent
    - During motions: cause error and `/mirror` will be cancelled.
    - During homing: cancel homing both master and slave
    - During servo mode: cause error and `/mirror` will be cancelled.
- Behaviors of during the mirror mode
    - When `/homing` sent
        - If `shareSensors` True, the slave will act same as it’s master.
        - If `shareSensors` False, both motors will do homing independently.
    - When `/setServoMode` sent
        - Both motors will be set to the servoMode. Then when `/setTargetPosition` sent, both motors will be move toward to the specified position. This motion could be different between motors, depends on the current position of both motors.

You suggested the slave’s sensors should behave normally even if `shareSensors` is true. But this behavior could be very complicated during homing. I think we should ignore the slave’s sensor triggers during homing.

And for me, a word of "mirror" sounds like the motion direction will be reversed. Probably "copy" is better? What you think about it as a native speaker?
Thanks!

**Reply by JulianOrteil** [reply]
@kanta Thanks for the more in depth analysis.

You are much more well tuned to how effective or complicated something like this will be. The only concern I have regarding the slave motor's sensors is, from a safety standpoint, if the home/limit sensor on the slave is activated and the motor continues past that point, then injuries could occur. If not injuries, then more than likely damage will. If you think it would be too complicated to implement my suggestion for the sensors, then it would be better to leave them both independent.

As for the name, "/copy" seems more fitting than "/mirror". May I throw out another suggestion of "/follow"? I know of a few other frameworks that use "/follow" for something like this suggestion.

My best,
Jules

**Reply by kanta** [reply]
@JulianOrteil Thank you for suggestions, "/follow" sounds nice! Let's take that word.

Some more thoughts about the sensors (sorry for too details!)
- If two motors are connected rigidly, `shareSensors` should be true and only the master should have sensors.
- If two motors are physically separated and have each own sensors, `shareSensors` should be false.
- The case `shareSensors` true and two motors have each own sensors, this sounds rare case but could be useful for certain situations.
  - I agree your thoughts about safety, always slave's sensors should stop slave motion.
  - In this case, the slave sensors will stop only slave, and the master sensors will stop both.
  - In a homing, if the master sensor reacts first, then slave will copy it and the home position will be wrong for the slave.
    - If the slave sensor reacts first, then should the slave proceed to `releaseSw` part independently? 
    - Or if the slave waits for the trigger from the master, many exceptions could happen. The releaseSw part of the slave will be possibly unfinished, then the home position will be wrong.
  
My conclusion so far is that if `shareSensors` is set to true, sensors from the slave should be ignored during the homing.
Does this condition work for your case?

Thanks!

**Reply by JulianOrteil** [reply]
@kanta Dont worry about how many details you provide, we are only hashing out the minutia, so the more the better!

> If the slave sensor reacts first, then should the slave proceed to releaseSw part independently?

I would think this should be fine. If both motors are so far off that this would cause an issue, then something else in the user's mechanism is wrong.

Id still like to see the slave sensors work as normal in homing, but if thats too complicated, then we should just put a warning in the documentation and address it later if needed.

Jules

**Reply by kanta** [reply]
@JulianOrteil Sorry I can't tell any update yet for this command. I expect to find a time to work for the `/follow` in the middle of December. Thank you for your patience.

**Comment by kanta** [comment]
@JulianOrteil One another strategy is preparing dedicated motion commands for these combined motions.
Something like 
`/combine/move 1 2 25600`
means executing following commands at the same time;
- `/move 1 25600`
- `/move 2 25600`

Or we can extend the command as
`/combine/move [num of axis] [axis 1]...[axis n] [steps]`

Those will be quite easier to implement so I would start with them as my first step.

**Reply by dakotahorstman** [reply]
I agree. I think that is a great suggestion.

**Reply by kanta** [reply]
@dakotahorstman Thank you for your comment! I have made `dev-follow` branch to work and updated briefly. 
https://github.com/ponoor/step-series-universal-firmware/tree/dev-follow

I haven't tested with a physical board yet but will try it soon.

**Reply by kanta** [reply]
@JulianOrteil I tested new commands and seems to be ok. Could you check them with your environment? The new firmware is available at the above branch.

New commands so far;
`/combine/move [axis# 1] [axis# 2] [steps]`
`/combine/goTo [axis# 1] [axis# 2] [target position]`
`/combine/goToDir [axis# 1] [axis# 2] [dir] [target position]`
`/combine/run [axis# 1] [axis# 2] [speed]`
`/combine/hardHiZ [axis# 1] [axis# 2]`
`/combine/hardStop [axis# 1] [axis# 2]`
`/combine/softHiZ [axis# 1] [axis# 2]`
`/combine/softStop [axis# 1] [axis# 2]`

**Reply by JulianOrteil** [reply]
@kanta I shall do so tomorrow morning and let you know my findings. I do not have an environment currently set up to test LIMIT and HOME sensors, so I will only be able to test if these commands help improve the sync issues between my motors.

Thanks for the quick work!
Jules
